### PR TITLE
[libc++] LWG3233 Broken requirements for shared_ptr converting constructors

### PR DESCRIPTION
**Pull request by huixie90**

### CODE REVIEW

File: libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared/util.smartptr.shared.const/types.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef TEST_STD_UTILITIES_MEMORY_UTIL_SMARTPTR_SHARED_CONST_TYPES_H
```
Comment (ldionne) [inline libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared/util.smartptr.shared.const/types.h:8]:
```suggestion
//===----------------------------------------------------------------------===//

#ifndef TEST_STD_UTILITIES_MEMORY_UTIL_SMARTPTR_SHARED_CONST_TYPES_H
```

**Review by ldionne — APPROVED** [review]

**Comment by alexfh** [comment]
This commit breaks the following code (https://gcc.godbolt.org/z/eoMMr6KjE):
```
    std::shared_ptr<int>(
      new int,
      [](auto pointer) { delete pointer; }
    );
```

Is this intended?

The code compiles fine with libstdc++ and with libc++ before this change.

**Comment by ldionne** [comment]
@alexfh I think this is correct behavior, however it is arguably not great. For example, MSVC behaves the same as libc++ after this patch: https://gcc.godbolt.org/z/xd5PdheeG

It looks like libstdc++ doesn't implement the resolution of this LWG issue yet and that's why they still accept this code.

Basically, what happens here is that the `nullptr_t` constructor overload of `std::shared_ptr` checks for the well-formedness of `deleter(ptr)`, where `ptr` has type `nullptr_t`. This instantiates the lambda and hard-errors inside the body of the lambda, because that's not an immediate context for SFINAE.

@CaseyCarter Was this an intended effect of the wording changes in http://wg21.link/LWG3233 ? CC @jwakely also

**Comment by jwakely** [comment]
No, I don't think that was considered. I think for the two overloads taking `nullptr_t` the constraint should be that `d((T*)nullptr)` is well-formed, since that's what the implementation actually does.

I think this needs a new LWG issue.

**Comment by jwakely** [comment]
N.B. I don't think this was caused by LWG 3233, the old wording said "this constructor shall not participate in overload resolution unless is_move_constructible_v<D> is true, the expression d(p) is well-formed" so the broken constraint was always present. It was even there in C++11, and violating it led to undefined behaviour (rather than being ill-formed). Libc++ just never enforced it until now.

**Comment by ldionne** [comment]
Email sent to LWG chair.

**Comment by jwakely** [comment]
> It was even there in C++11, and violating it led to undefined behaviour (rather than being ill-formed).

I'm wrong about this - it was fine in C++11, because we didn't try to test `d(p)` during overload resolution, so didn't get a hard error outside the immediate context. The problem is caused by checking the `d(p)` expression during overload resolution, when considering the `shared_ptr(nullptr_t, D)` ctor, which is never actually going to be selected for this call.


> I think for the two overloads taking `nullptr_t` the constraint should be that `d((T*)nullptr)` is well-formed, since that's what the implementation actually does.

And I was wrong about this too! As pointed out by Peter Dimov. The implementation really does do `d(nullptr)` if the `shared_ptr(nullptr_t, D)` constructor is called. But the problematic example above doesn't actually call that constructor, it just evaluates its constraints during overload resolution.

**Comment by ldionne** [comment]
I read the discussion on the LWG reflector. I see the proposed resolution isn't actually correct.

Given this, and since it looks like this is going to require additional work both in WG21 and in the implementations, @alexfh I think I will have to revert my workaround for this, and you folks will have to change your code at least for the time being. I'm sorry if this ends up flip-flopping once WG21 solves this issue, but for the time being we have to be Standards conforming and we are standards conforming in the state where your code is broken. It's unfortunate but I think it's the only reasonable state to be in for libc++.

**Comment by jwakely** [comment]
It was noted on the reflector that using `[](auto* pointer) { delete pointer; }` works fine, because that deleter gives a substitution error when the `d(nullptr)` constraint is tested, instead of a hard error outside the immediate context.

**Comment by bgra8** [comment]
@ldionne this also stopped working: https://gcc.godbolt.org/z/K8ncoP4Pv, is that expected??

**Comment by ldionne** [comment]
> @ldionne this also stopped working: https://gcc.godbolt.org/z/K8ncoP4Pv, is that expected??

I don't know yet, I have not had time to look into it and I won't until next week.

For this reason, I am going to temporarily revert this patch since it seems to cause issues and we don't know yet whether we have a bug in there or not. It's also a small and self-contained patch so easy to revert/reapply. I will investigate this next week and we can re-apply it.

**Comment by ldionne** [comment]
We should be in a stable state right now, i.e. back to how things were before we tried fixing LWG3233. I just spoke to @huixie90 who agreed to pick this up again and look into whether https://gcc.godbolt.org/z/K8ncoP4Pv is valid code. We'll ping you when a new PR to fix this LWG issue is ready.

**Comment by alexfh** [comment]
Thank you @ldionne! We've done some cleanup in our code to prepare at least for relanding of LWG3233. I suppose, we shouldn't see any more fallout unless larger changes are made in this area.

But I think, stability of libc++ interface is valuable to other users as well. So figuring out how the end state should look like would be nice before doing the change.

**Pull request merged**